### PR TITLE
Complete CRC proposal withdraw transaction related logic

### PR DESCRIPTION
**Pull request by Houshoupei84**

### CODE REVIEW

File: cr/state/proposalmanager.go [hunk 1]
```diff
@@ -351,6 +374,21 @@ func (p *ProposalManager) proposalReview(tx *types.Transaction,
...
+func (p *ProposalManager) proposalWithdraw(tx *types.Transaction,
+	height uint32, history *utils.History) {
+	proposalWithdraw := tx.Payload.(*payload.CRCProposalWithdraw)
+	proposalState := p.getProposal(proposalWithdraw.ProposalHash)
+	if proposalState == nil {
+		return
+	}
+
+	history.Append(height, func() {
+		proposalState.CurrentWithdrawalStage++
```
Comment (RainFallsSilent) [inline cr/state/proposalmanager.go]: should consider that withdraw may contain more than one stage

File: blockchain/txvalidator.go [hunk 2]
```diff
@@ -796,10 +850,24 @@ func (b *BlockChain) checkTransactionFee(tx *Transaction, references map[*Input]
 		inputValue += output.Value
 	}
 	if inputValue < b.chainParams.MinTransactionFee+outputValue {
```
Comment (RainFallsSilent) [inline blockchain/txvalidator.go]: getTransactionFee has check,  better to put the check out of this function ?

File: blockchain/txvalidator.go [hunk 3]
```diff
@@ -1595,6 +1672,55 @@ func (b *BlockChain) checkCRCProposalReviewTransaction(txn *Transaction,
...
+	if proposalState.Status != crstate.VoterAgreed {
+		return errors.New("proposal status is not VoterAgreed")
+	}
+	if withdrawPayload.Stage <= proposalState.CurrentWithdrawalStage {
+		return errors.New("CRCProposalWithdraw Stage wrong too small")
+	}
+	if withdrawPayload.Stage >= proposalState.CurrentStage {
+		return errors.New("CRCProposalWithdraw Stage wrong too big")
+	}
+	if !bytes.Equal(proposalState.Proposal.SponsorPublicKey, withdrawPayload.SponsorPublicKey) {
```
Comment (RainFallsSilent) [inline blockchain/txvalidator.go]: should use proposalStat.ProposalLeader, sponsor publicKey may have changed.

File: cr/state/proposalmanager.go [hunk 4]
```diff
@@ -209,6 +209,25 @@ func (p *ProposalManager) transferRegisteredState(proposal *ProposalState,
 	}
 }
 
+func (p *ProposalManager) CanWithdrawalAmount(hash common.Uint256) common.Fixed64 {
```
Comment (raindust) [inline cr/state/proposalmanager.go]: rename to "AvailableWithdrawalAmount" would be better

File: core/types/payload/crcproposalwithdraw.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	// Hash of the proposal to withdrawal ela
+	ProposalHash common.Uint256
+
+	// Public key of sponsor.
+	SponsorPublicKey []byte
+
+	// Which stage of money to take
+	Stage uint8
+
+	//fee count
```
Comment (raindust) [inline core/types/payload/crcproposalwithdraw.go]: comment should be "fee amount"

File: blockchain/txvalidator.go [hunk 6]
```diff
@@ -1596,6 +1679,57 @@ func (b *BlockChain) checkCRCProposalReviewTransaction(txn *Transaction,
 	return checkCRTransactionSignature(crcProposalReview.Sign, crMember.Info.Code,
 		signedBuf.Bytes())
 }
+func GetCode(publicKey []byte) []byte {
+	pk, _ := crypto.DecodePoint(publicKey)
+	redeemScript, _ := contract.CreateStandardRedeemScript(pk)
+	return redeemScript
+}
```
Comment (raindust) [inline blockchain/txvalidator.go]: new line between two methods

File: blockchain/txvalidator.go [hunk 7]
```diff
@@ -603,6 +637,13 @@ func (b *BlockChain) checkTransactionOutput(blockHeight uint32,
 	if len(txn.Outputs) < 1 {
 		return errors.New("transaction has no outputs")
 	}
+
+	if txn.IsCRCProposalWithdrawTx() {
+		err := b.checkCRCProposalWithdrawOutput(txn)
+		if err != nil {
+			return err
```
Comment (peterwillcn) [inline blockchain/txvalidator.go]: Use return elaerr.Simple(,)

**Pull request merged**